src/System.Text.Primitives/System/Text/Encoding/Utf8/Utf8Encoder.cs

```diff
@@ -290,53 +300,119 @@ internal static bool TryComputeEncodedBytes(ReadOnlySpan<uint> utf32, out int by
+            int i = 0;
+            char codePoint = '\0';
 
-            for (int i = 0; i < availableCharacters; i++)
+            unsafe
             {
-                var ch = utf16[i];
+                var utf16Ref = Unsafe.AsPointer(ref utf16.DangerousGetPinnableReference());
+                var utf8Ref = Unsafe.AsPointer(ref utf8.DangerousGetPinnableReference());
 
-                if ((ushort)ch <= 0x7f) // this if block just optimizes for ascii
+                var utf16CharRef = (char*)utf16Ref;
+                var utf8ByteRef = (byte*)utf8Ref;
+
+                i = 0;
+
+                ascii:
+                for (; i < inputCharacters; i++)
                 {
-                    if (bufferLength - bytesWritten < 1)
-                        return false;
+                    codePoint = *(utf16CharRef + i);
 
-                    utf8[bytesWritten++] = (byte)ch;
-                    charactersConsumed++;
+                    if (codePoint > Utf8OneByteLastCodePoint)
+                    {
+                        goto non_ascii;
+                    }
+                    if (bytesWrittenTemp + 1 > outputBytesAvailable)
+                    {
+                        goto need_more;
+                    }
+                    *(utf8ByteRef + bytesWrittenTemp) = (byte)codePoint;
+                    bytesWrittenTemp++;
                 }
-                else
-                {
-                    uint codePoint;
-                    int consumed;
 
-                    if (!char.IsSurrogate(ch))
+                non_ascii:
+                for (; i < inputCharacters; i++)
+                {
+                    codePoint = *(utf16CharRef + i);
+                    if (codePoint <= Utf8OneByteLastCodePoint)
+                    {
+                        goto ascii;
+                    }
+                    else if (codePoint <= Utf8TwoBytesLastCodePoint)
+                    {
+                        if (bytesWrittenTemp + 2 > outputBytesAvailable)
+                        {
+                            goto need_more;
+                        }
+                        *(utf8ByteRef + bytesWrittenTemp) = (byte)(((codePoint >> 6) & b0001_1111U) | b1100_0000U);
+                        bytesWrittenTemp++;
+                        *(utf8ByteRef + bytesWrittenTemp) = (byte)(((codePoint) & b0011_1111U) | b1000_0000U);
+                        bytesWrittenTemp++;
+                    }
+                    else if (codePoint >= Utf16HighSurrogateFirstCodePoint && codePoint <= Utf16HighSurrogateLastCodePoint)
+                    {
+                        if (bytesWrittenTemp + 4 > outputBytesAvailable)
+                        {
+                            goto need_more;
+                        }
+
+                        if (++i >= inputCharacters)
+                            throw new ArgumentException("Invalid surrogate pair.", nameof(utf16));
+
+                        var lowSurrogate = *(utf16CharRef + i);
```

Comment:
Nit: Coding guidelines allow use of `var` only if the type shows up the right hand side. It is not the case here.